File: scripts/enum/s3.js
'use strict';

const {
  S3Client,
  ListBucketsCommand,
  GetBucketLocationCommand,
  GetBucketPolicyCommand,
  GetPublicAccessBlockCommand,
  GetBucketVersioningCommand,
  GetBucketEncryptionCommand,
  GetBucketLoggingCommand,
  GetBucketAclCommand,
} = require('@aws-sdk/client-s3');
// NO STS import — base-enum handles it

const { withRetry, createLogger } = require('../lib');
const { baseEnum } = require('../lib/base-enum');

// --- Helpers ---

/**
 * Normalizes the LocationConstraint response to a region string.
 * null/empty means us-east-1 (S3 legacy behavior).
 */
function normalizeBucketRegion(locationConstraint) {
  if (!locationConstraint || locationConstraint === '') return 'us-east-1';
  if (locationConstraint === 'EU') return 'eu-west-1';
  return locationConstraint;
}

/**
 * Safely calls an S3 API, returning null for expected "not configured" errors.
 */
async function safeGetBucketConfig(fn, notFoundCodes) {
  try {
    return await withRetry(fn);
  } catch (err) {
    const code = err.name || err.Code || '';
    if (notFoundCodes.includes(code)) {
      return null;
    }
    throw err;
  }
}

// --- Findings generation ---

function generateFindings(bucket) {
  const findings = [];

  // Public access check via ACL
  if (bucket.acl_grants) {
    const publicGrants = bucket.acl_grants.filter(
      (g) => g.grantee_uri && (
        g.grantee_uri.includes('AllUsers') ||
        g.grantee_uri.includes('AuthenticatedUsers')
      )
    );
    if (publicGrants.length > 0) {
      findings.push({
        type: 'public_acl',
        severity: 'high',
        detail: `Bucket has ${publicGrants.length} public ACL grant(s)`,
      });
    }
  }

  // Public access block not set
  if (!bucket.public_access_block) {
    findings.push({
      type: 'no_public_access_block',
      severity: 'medium',
      detail: 'No public access block configuration set',
    });
  } else {
    const pab = bucket.public_access_block;
    if (!pab.BlockPublicAcls || !pab.IgnorePublicAcls || !pab.BlockPublicPolicy || !pab.RestrictPublicBuckets) {
      findings.push({
        type: 'partial_public_access_block',
        severity: 'medium',
        detail: 'Public access block is not fully restrictive',
      });
    }
  }

  // No encryption
  if (!bucket.encryption) {
    findings.push({
      type: 'no_encryption',
      severity: 'medium',
      detail: 'No default server-side encryption configured',
    });
  }

  // No versioning
  if (!bucket.versioning || bucket.versioning.Status !== 'Enabled') {
    findings.push({
      type: 'no_versioning',
      severity: 'low',
      detail: 'Bucket versioning is not enabled',
    });
  }

  // Overly permissive policy (wildcard principal)
  if (bucket.policy) {
    try {
      const doc = JSON.parse(bucket.policy);
      const statements = Array.isArray(doc.Statement) ? doc.Statement : [];
      for (const stmt of statements) {
        if (stmt.Effect !== 'Allow') continue;
        const principal = stmt.Principal;
        if (principal === '*' || (principal && principal.AWS === '*')) {
          findings.push({
            type: 'public_policy',
            severity: 'high',
            detail: 'Bucket policy allows wildcard principal',
          });
          break;
        }
      }
    } catch {
      // unparseable policy — skip
    }
  }

  return findings;
}

// --- Run (dependency-injectable) ---

async function run(opts = {}) {
  const runDir = opts.runDir;
  const region = opts.region;
  const accountId = opts.accountId;

  const s3Client = opts.clients?.s3 ?? new S3Client({ region: 'us-east-1' });

  const logger = opts.logger || createLogger(runDir, 's3');
  logger.log('info', 'S3_Enumeration_Start', { region });

  // ListBuckets is global — use the provided s3 client (defaults to us-east-1)
  let allBuckets;
  try {
    logger.log('api_call', 'ListBuckets', { service: 's3' });
    const resp = await withRetry(() => s3Client.send(new ListBucketsCommand({})));
    allBuckets = resp.Buckets || [];
  } catch (err) {
    logger.log('error', 'ListBuckets', { error: err.message });
    await logger.flush();
    throw new Error(`ListBuckets failed: ${err.message}`);
  }

  // Discover bucket regions and filter to requested region
  const findings = [];
  let status = 'complete';
  const errors = [];

  for (const bucket of allBuckets) {
    const bucketName = bucket.Name;
    let bucketRegion;

    try {
      logger.log('api_call', 'GetBucketLocation', { bucket: bucketName });
      const locResp = await withRetry(() =>
        s3Client.send(new GetBucketLocationCommand({ Bucket: bucketName }))
      );
      bucketRegion = normalizeBucketRegion(locResp.LocationConstraint);
    } catch (err) {
      // Can't determine region — skip bucket
      errors.push({ resource: bucketName, error: err.message });
      logger.log('warning', 'GetBucketLocation', { bucket: bucketName, error: err.message });
      continue;
    }

    // Only enumerate buckets in the requested region
    if (bucketRegion !== region) continue;

    // Use the provided client for regional requests too
    const regionalClient = s3Client;

    const bucketFinding = {
      resource_type: 's3_bucket',
      resource_id: bucketName,
      arn: `arn:aws:s3:::${bucketName}`,
      region: bucketRegion,
      policy: null,
      public_access_block: null,
      versioning: null,
      encryption: null,
      logging: null,
      acl_grants: null,
      findings: [],
    };

    // GetBucketPolicy
    try {
      logger.log('api_call', 'GetBucketPolicy', { bucket: bucketName });
      const policyResp = await safeGetBucketConfig(
        () => regionalClient.send(new GetBucketPolicyCommand({ Bucket: bucketName })),
        ['NoSuchBucketPolicy']
      );
      bucketFinding.policy = policyResp ? policyResp.Policy : null;
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetBucketPolicy: ${err.message}` });
      logger.log('warning', 'GetBucketPolicy', { bucket: bucketName, error: err.message });
    }

    // GetPublicAccessBlock
    try {
      logger.log('api_call', 'GetPublicAccessBlock', { bucket: bucketName });
      const pabResp = await safeGetBucketConfig(
        () => regionalClient.send(new GetPublicAccessBlockCommand({ Bucket: bucketName })),
        ['NoSuchPublicAccessBlockConfiguration']
      );
      bucketFinding.public_access_block = pabResp ? pabResp.PublicAccessBlockConfiguration : null;
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetPublicAccessBlock: ${err.message}` });
      logger.log('warning', 'GetPublicAccessBlock', { bucket: bucketName, error: err.message });
    }

    // GetBucketVersioning
    try {
      logger.log('api_call', 'GetBucketVersioning', { bucket: bucketName });
      const verResp = await withRetry(() =>
        regionalClient.send(new GetBucketVersioningCommand({ Bucket: bucketName }))
      );
      bucketFinding.versioning = { Status: verResp.Status || null, MFADelete: verResp.MFADelete || null };
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetBucketVersioning: ${err.message}` });
      logger.log('warning', 'GetBucketVersioning', { bucket: bucketName, error: err.message });
    }

    // GetBucketEncryption
    try {
      logger.log('api_call', 'GetBucketEncryption', { bucket: bucketName });
      const encResp = await safeGetBucketConfig(
        () => regionalClient.send(new GetBucketEncryptionCommand({ Bucket: bucketName })),
        ['ServerSideEncryptionConfigurationNotFoundError']
      );
      bucketFinding.encryption = encResp
        ? encResp.ServerSideEncryptionConfiguration
        : null;
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetBucketEncryption: ${err.message}` });
      logger.log('warning', 'GetBucketEncryption', { bucket: bucketName, error: err.message });
    }

    // GetBucketLogging
    try {
      logger.log('api_call', 'GetBucketLogging', { bucket: bucketName });
      const logResp = await withRetry(() =>
        regionalClient.send(new GetBucketLoggingCommand({ Bucket: bucketName }))
      );
      bucketFinding.logging = logResp.LoggingEnabled || null;
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetBucketLogging: ${err.message}` });
      logger.log('warning', 'GetBucketLogging', { bucket: bucketName, error: err.message });
    }

    // GetBucketAcl
    try {
      logger.log('api_call', 'GetBucketAcl', { bucket: bucketName });
      const aclResp = await withRetry(() =>
        regionalClient.send(new GetBucketAclCommand({ Bucket: bucketName }))
      );
      bucketFinding.acl_grants = (aclResp.Grants || []).map((g) => ({
        grantee_type: g.Grantee?.Type || null,
        grantee_id: g.Grantee?.ID || null,
        grantee_uri: g.Grantee?.URI || null,
        permission: g.Permission || null,
      }));
    } catch (err) {
      errors.push({ resource: bucketName, error: `GetBucketAcl: ${err.message}` });
      logger.log('warning', 'GetBucketAcl', { bucket: bucketName, error: err.message });
    }

    // Generate findings
    bucketFinding.findings = generateFindings(bucketFinding);
    findings.push(bucketFinding);
  }

  if (errors.length > 0) status = 'partial';

  await logger.flush();
  return { findings, status };
}

if (require.main === module) {
  baseEnum({ module: 's3', run });
}

module.exports = { run };

File: scope/enumerators/s3.py
from __future__ import annotations

import json
from typing import Any, Callable

from scope.core.coverage import CoverageTracker
from scope.core.models import ModuleEnvelope
from scope.enumerators.resource_policy import normalize_resource_policy


PRIMARY_CHECKS = ["list_buckets"]
REQUIRED_CHECKS = ["bucket_policy", "bucket_acl", "bucket_public_access_block"]


def run(factory: Any, region: str) -> ModuleEnvelope:
    tracker = CoverageTracker(primary=PRIMARY_CHECKS, required=REQUIRED_CHECKS)
    s3 = factory.client("s3")
    findings: list[dict[str, Any]] = []

    try:
        buckets = (s3.list_buckets().get("Buckets") or [])
        tracker.record_ok("list_buckets", resource="buckets")
    except Exception as err:
        tracker.record_module_failure("list_buckets", operation="s3.ListBuckets", err=err)
        coverage, errors = tracker.to_envelope_fields()
        return ModuleEnvelope(
            module="s3",
            account_id=factory.account_id,
            region=region,
            status=tracker.derive_status(),
            resources=[],
            coverage=coverage,
            errors=errors,
        )

    for bucket in buckets:
        bucket_name = bucket.get("Name")
        bucket_arn = f"arn:aws:s3:::{bucket_name}"
        try:
            location = s3.get_bucket_location(Bucket=bucket_name)
            bucket_region = _normalize_bucket_region(location.get("LocationConstraint"))
        except Exception as err:
            tracker.record_skipped("bucket_location", resource=bucket_arn, reason=_error_code(err))
            continue

        if region != "global" and bucket_region != region:
            continue

        finding = {
            "resource_type": "s3_bucket",
            "resource_id": bucket_name,
            "arn": bucket_arn,
            "region": bucket_region,
            "policy": None,
            "policy_status": "absent",
            "public_access_block": None,
            "public_access_block_status": "absent",
            "versioning": None,
            "versioning_status": "absent",
            "encryption": None,
            "encryption_status": "absent",
            "logging": None,
            "logging_status": "absent",
            "acl_grants": None,
            "acl_status": "absent",
            "findings": [],
        }

        try:
            response = _safe_get(lambda: s3.get_bucket_policy(Bucket=bucket_name), {"NoSuchBucketPolicy"})
            if response:
                finding["policy"] = response.get("Policy")
                normalized_policy = normalize_resource_policy(response.get("Policy"), account_id=factory.account_id)
                if normalized_policy:
                    finding["resource_policy"] = {"principals": normalized_policy["principals"]}
                    finding["resource_policy_document"] = normalized_policy["document"]
                    finding["resource_policy_statements"] = normalized_policy["statements"]
                finding["policy_status"] = "present"
            else:
                finding["policy_status"] = "absent"
            tracker.record_ok("bucket_policy", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["policy_status"] = _classify_status(code)
            tracker.record_resource_failure("bucket_policy", resource=bucket_arn, operation="s3.GetBucketPolicy", err=err)

        try:
            response = _safe_get(
                lambda: s3.get_public_access_block(Bucket=bucket_name),
                {"NoSuchPublicAccessBlockConfiguration"},
            )
            if response:
                finding["public_access_block"] = response.get("PublicAccessBlockConfiguration")
                finding["public_access_block_status"] = "present"
            else:
                finding["public_access_block_status"] = "absent"
            tracker.record_ok("bucket_public_access_block", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["public_access_block_status"] = _classify_status(code)
            tracker.record_resource_failure(
                "bucket_public_access_block",
                resource=bucket_arn,
                operation="s3.GetPublicAccessBlock",
                err=err,
            )

        try:
            response = s3.get_bucket_versioning(Bucket=bucket_name)
            finding["versioning"] = {
                "Status": response.get("Status"),
                "MFADelete": response.get("MFADelete"),
            }
            finding["versioning_status"] = "present" if response.get("Status") else "absent"
            tracker.record_ok("bucket_versioning", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["versioning_status"] = _classify_status(code)
            tracker.record_skipped("bucket_versioning", resource=bucket_arn, reason=_skip_reason(code))

        try:
            response = _safe_get(
                lambda: s3.get_bucket_encryption(Bucket=bucket_name),
                {"ServerSideEncryptionConfigurationNotFoundError"},
            )
            if response:
                finding["encryption"] = response.get("ServerSideEncryptionConfiguration")
                finding["encryption_status"] = "present"
            else:
                finding["encryption_status"] = "absent"
            tracker.record_ok("bucket_encryption", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["encryption_status"] = _classify_status(code)
            tracker.record_skipped("bucket_encryption", resource=bucket_arn, reason=_skip_reason(code))

        try:
            response = s3.get_bucket_logging(Bucket=bucket_name)
            finding["logging"] = response.get("LoggingEnabled")
            finding["logging_status"] = "present" if response.get("LoggingEnabled") else "absent"
            tracker.record_ok("bucket_logging", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["logging_status"] = _classify_status(code)
            tracker.record_skipped("bucket_logging", resource=bucket_arn, reason=_skip_reason(code))

        try:
            response = s3.get_bucket_notification_configuration(Bucket=bucket_name)
            notification = _notification_configuration(response)
            if notification:
                finding["notification_configuration"] = notification
                finding["notification_configuration_status"] = "present"
            else:
                finding["notification_configuration_status"] = "absent"
            tracker.record_ok("bucket_notification_configuration", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["notification_configuration_status"] = _classify_status(code)
            tracker.record_skipped("bucket_notification_configuration", resource=bucket_arn, reason=_skip_reason(code))

        try:
            response = s3.get_bucket_acl(Bucket=bucket_name)
            finding["acl_grants"] = [
                {
                    "grantee_type": (grant.get("Grantee") or {}).get("Type"),
                    "grantee_id": (grant.get("Grantee") or {}).get("ID"),
                    "grantee_uri": (grant.get("Grantee") or {}).get("URI"),
                    "permission": grant.get("Permission"),
                }
                for grant in response.get("Grants") or []
            ]
            finding["acl_status"] = "present"
            tracker.record_ok("bucket_acl", resource=bucket_arn)
        except Exception as err:
            code = _error_code(err)
            finding["acl_status"] = _classify_status(code)
            tracker.record_resource_failure("bucket_acl", resource=bucket_arn, operation="s3.GetBucketAcl", err=err)

        finding["findings"] = _generate_findings(finding)
        findings.append(finding)

    coverage, errors = tracker.to_envelope_fields()
    return ModuleEnvelope(
        module="s3",
        account_id=factory.account_id,
        region=region,
        status=tracker.derive_status(),
        resources=findings,
        coverage=coverage,
        errors=errors,
    )


def _normalize_bucket_region(location_constraint: str | None) -> str:
    if not location_constraint:
        return "us-east-1"
    if location_constraint == "EU":
        return "eu-west-1"
    return location_constraint


def _safe_get(call: Callable[[], dict[str, Any]], not_found_codes: set[str]) -> dict[str, Any] | None:
    try:
        return call()
    except Exception as err:
        if _error_code(err) in not_found_codes:
            return None
        raise


def _generate_findings(bucket: dict[str, Any]) -> list[dict[str, str]]:
    findings: list[dict[str, str]] = []
    public_grants = [
        grant
        for grant in bucket.get("acl_grants") or []
        if grant.get("grantee_uri")
        and ("AllUsers" in grant["grantee_uri"] or "AuthenticatedUsers" in grant["grantee_uri"])
    ]
    if public_grants:
        findings.append(
            {
                "type": "public_acl",
                "severity": "high",
                "detail": f"Bucket has {len(public_grants)} public ACL grant(s)",
            }
        )

    pab = bucket.get("public_access_block")
    if not pab:
        findings.append(
            {
                "type": "no_public_access_block",
                "severity": "medium",
                "detail": "No public access block configuration set",
            }
        )
    elif not (
        pab.get("BlockPublicAcls")
        and pab.get("IgnorePublicAcls")
        and pab.get("BlockPublicPolicy")
        and pab.get("RestrictPublicBuckets")
    ):
        findings.append(
            {
                "type": "partial_public_access_block",
                "severity": "medium",
                "detail": "Public access block is not fully restrictive",
            }
        )

    if not bucket.get("encryption"):
        findings.append({"type": "no_encryption", "severity": "medium", "detail": "No default server-side encryption configured"})

    versioning = bucket.get("versioning")
    if not versioning or versioning.get("Status") != "Enabled":
        findings.append({"type": "no_versioning", "severity": "low", "detail": "Bucket versioning is not enabled"})

    if _has_public_policy(bucket.get("policy")):
        findings.append(
            {
                "type": "public_policy",
                "severity": "high",
                "detail": "Bucket policy allows wildcard principal",
            }
        )

    return findings


def _has_public_policy(policy: str | None) -> bool:
    if not policy:
        return False
    try:
        document = json.loads(policy)
    except json.JSONDecodeError:
        return False
    statements = document.get("Statement", []) if isinstance(document, dict) else []
    if isinstance(statements, dict):
        statements = [statements]
    for statement in statements:
        if not isinstance(statement, dict) or statement.get("Effect") != "Allow":
            continue
        principal = statement.get("Principal")
        if principal == "*":
            return True
        if isinstance(principal, dict) and principal.get("AWS") == "*":
            return True
    return False


def _notification_configuration(response: dict[str, Any]) -> dict[str, Any]:
    keys = (
        "LambdaFunctionConfigurations",
        "TopicConfigurations",
        "QueueConfigurations",
        "EventBridgeConfiguration",
    )
    return {key: response[key] for key in keys if response.get(key)}


def _classify_status(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else "error"


def _skip_reason(code: str) -> str:
    return "access_denied" if code in {"AccessDenied", "AccessDeniedException"} else code


def _error_code(err: Exception) -> str:
    response = getattr(err, "response", None)
    if isinstance(response, dict):
        error = response.get("Error")
        if isinstance(error, dict) and error.get("Code"):
            return str(error["Code"])
    code = getattr(err, "Code", None)
    if code:
        return str(code)
    return err.__class__.__name__ or "Error"
